Packs/Iboss/Integrations/Iboss/Iboss.py

```diff
@@ -0,0 +1,1532 @@
+            "startPort": start_port,
+            "timedUrlExpiresInMinutes": time_url_expires_in_minutes,
+        }
+
+        result = self._http_request(
+            method="put",
+            url_suffix="/json/controls/blockList",
+            json_data=params,
+            params={"currentPolicyBeingEdited": current_policy_being_edited},
+            error_handler=self._list_error_handler,
+            ok_codes=(200, 201, 202, 204, 422)
+        )
+        return result
+
+    @set_auth
+    def remove_entity_from_block_list(self, url, current_policy_being_edited=1, direction=2, end_port=0, start_port=0):
+        params = {
+            "url": url,
+            "currentPolicyBeingEdited": current_policy_being_edited,
+            "direction": direction,
+            "endPort": end_port,
+            "startPort": start_port,
+        }
+
+        result = self._http_request(
+            method="delete",
+            url_suffix="/json/controls/blockList",
+            params=params,
+            ok_codes=(200, 201, 202, 204, 422),
+            error_handler=self._remove_from_list_error_handler,
+        )
+        return result
+
+    @set_auth
+    def add_entity_to_allow_list(self, url, current_policy_being_edited=1, allow_keyword=0,
+                                 direction=2, end_port=0, apply_global=0, is_regex=0, is_timed_url=0, note="",
+                                 priority=0, start_port=0, time_url_expires_in_minutes=0):
+        params = {
+            "url": url,
+            "allow_keyword": allow_keyword,
+            "direction": direction,
+            "endPort": end_port,
+            "global": apply_global,
+            "isRegex": is_regex,
+            "isTimedUrl": is_timed_url,
+            "note": note,
+            "priority": priority,
+            "startPort": start_port,
+            "timedUrlExpiresInMinutes": time_url_expires_in_minutes,
+        }
+
+        result = self._http_request(
+            method="put",
+            url_suffix="/json/controls/allowList",
+            json_data=params,
+            params={"currentPolicyBeingEdited": current_policy_being_edited},
+            error_handler=self._list_error_handler,
+            ok_codes=(200, 201, 202, 204, 422)
+        )
+        return result
+
+    @set_auth
+    def remove_entity_from_allow_list(self, url, current_policy_being_edited=1, direction=2, end_port=0, start_port=0):
+        params = {
+            "url": url,
+            "currentPolicyBeingEdited": current_policy_being_edited,
+            "direction": direction,
+            "endPort": end_port,
+            "startPort": start_port,
+        }
+
+        result = self._http_request(
+            method="delete",
+            url_suffix="/json/controls/allowList",
+            params=params,
+            ok_codes=(200, 201, 202, 204, 422),
+            error_handler=self._remove_from_list_error_handler
+        )
+        return result
+
+    @set_auth
```

Comment:
Yes! I should have added a docstring to explain this (and can do so if you approve this message). Whenever a user first authenticates to iboss there are a few subsequent API calls made (this is the standard auth method)
1. Get the cloud token
2. Determine the primary gateway to interact with (think of this as a separate API/server behind the main server)
3. Authenticate with that gateway
4. Establish session credentials with the that gateway

If I set the authentication every time, that will be multiple requests for every command. What I attempt to do is leverage the Demisto cache to store the tokens and expiration times, so on subsequent calls the integration can leverage the cache and not make these requests. On L184 we set the authentication for an expired token.

That said, there are times that the iboss server may revoke a token prior to expiration. This can occur due to user session conflict and some other edge cases. For such cases, the API will return an unauthorized message or a "Please login" message depending on the API endpoint. L188 makes a check for these cases and the `set_auth` wrapper will attempt to authenticate and try the failed request one more time. If there is truly an unauth issue, then the request will still fail after the second attempt, but when working as intended the second request attempt carries onwards.

Ultimately I made this a wrapper function to emulate something like the retry decorators that a library like `tenacity` provides. In doing so, if a command fails for auth due to session conflict or token expiration, the user will not be presented with a failed login but the expected response after the reauthentication automatically occurs.

Reply:
Thank you for the detailed explanation. 
It seems good, I'm okay with that as long as you tested it and the described logic works as expected. 
You can also add an explanation about it to the docstring if you like to :)